text/3204-references-alignment-niches.md

```diff
@@ -0,0 +1,144 @@
+- Feature Name: `references_alignment_niches`
+- Start Date: 2021-12-06
+- RFC PR: [rust-lang/rfcs#3204](https://github.com/rust-lang/rfcs/pull/3204)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add new "alignment niches" to references (`&'a T`, `&'a mut T`), `Vec<T>`, and all smart pointer types, allowing better packing for `enum`s containing these types.
+
+# Motivation
+[motivation]: #motivation
+
+Consider the following `enum`:
+
+```rs
+enum E<'a> {
+    A(&'a u16),
+    B,
+    C,
+}
+```
+
+Currently (on 64 bit targets), `mem::size_of::<E>() == 16`:
+  - 8 bytes for the `&'a u16`;
+  - 1 byte for the discriminant;
+  - 7 bytes of padding.
+
+However, this is suboptimal; because `&'a u16` must be well-aligned (in particular, 2-aligned), there are unused bit patterns that can be exploited to pack the `enum` into a single `usize`:
+- `E::A(&'a u16)` stores the reference directly;
+- `E::B` is stored as the bit pattern `0b0`;
+- `E::C` is stored as the bit pattern `0b1`.
+
+This RFC aims to introduces new niches to references types and smart pointer types, so that such `enum` optimizations become possible.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+"Pointer-like" containers (`&T`, `&mut T`, etc) gain new niches, corresponding to the scalar values strictly below `align(T)` and strictly above `align(T).wrapping_neg()`. For 1-aligned types, this is equivalent to the existing null niche.
+
+The full list of affected types is: `&T, &mut T, Box<T>, Arc<T>, Rc<T>, {sync, rc}::Weak<T>, Vec<T>`.
+
+
+This has the consequence that `enum`s of the form:
+```rs
+enum E {
+  A(&T or &mut T or ...),
+  B,
+  C,
+  // ...
+}
+```
+where the total number of variants is less than or equal to `2*align(T)` may be optimized by the compiler to fit in the space of a single reference.
```

Comment:
> `2*align(T)`

Should this be `2 ** align(T)` ?

Reply:
No, this is correct; this RFC only provides for values 0–align(T) and (max-align(T))–max as niches.

The current niche architecture in the compiler only provides for valid_scalar_range_start and valid_scalar_range_end, not for modulo validity.

Or IOW, a value with the low bits set is an invalid reference, but not used as a niche per just this RFC. (Using such values as niches additionally increases the cost of extracting such niched values.)

Reply:
Gotya, thanks for the explanation